sql/pgwire/encoding.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Ben Darnell
+
+package pgwire
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"io"
+	"unsafe"
+
+	"github.com/cockroachdb/cockroach/util"
+)
+
+const maxMessageSize = 1 << 24
+
+var _ bufferedReader = &bufio.Reader{}
+var _ bufferedReader = &bytes.Buffer{}
+
+type bufferedReader interface {
+	io.Reader
+	ReadString(delim byte) (string, error)
+	ReadByte() (byte, error)
+}
+
+type readBuffer struct {
+	msg []byte
+	tmp [4]byte
+}
+
+// reset sets b.msg to exactly size, attempting to use spare capacity
+// at the end of the existing slice when possible and allocating a new
+// slice when necessary.
+func (b *readBuffer) reset(size int) {
+	if b.msg != nil {
+		b.msg = b.msg[len(b.msg):]
```

Comment:
@tamird re: reviewable. Doesn't this mean that we never re-use anything of the existing buffer? So we'd alloc 4096 initially, use it up across multiple calls to `reset(.)` and then allocate another (presumably) 4096.